dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/tools/BeanMethodHandler.java

```diff
@@ -112,11 +121,61 @@ protected ToolResponse createSuccessfulResponse(Object result) {
             return ToolResponse.success(s);
         } else if (method.isStructuredContent()) {
             return new ToolResponse(false, List.of(new TextContent(jsonb.toJson(result))), result, null);
+        } else {
+            ToolArgumentsImpl toolArgumentsImpl = (ToolArgumentsImpl) toolArgs;
+            return encodeResult(result, toolArgumentsImpl.encoderRegistry());
+        }
+    }
+
+    private ToolResponse encodeResult(Object result, EncoderRegistry encoderRegistry) {
+        Class<?> resultType = result.getClass();
+        if (result instanceof List<?> list && !list.isEmpty()) {
+            resultType = list.get(0).getClass();
+        }
+        Optional<Encoder<?, ?>> encoder = encoderRegistry.findEncoder(resultType);
+
+        if (encoder.isPresent()) {
+            return encodeResultWithEncoder(result, encoder.get(), jsonb);
         } else {
             return ToolResponse.success(Objects.toString(result));
         }
     }
 
+    @SuppressWarnings("unchecked")
+    public ToolResponse encodeResultWithEncoder(Object result, Encoder<?, ?> encoder, Jsonb jsonb) {
+        try {
+            if (encoder instanceof ToolResponseEncoder) {
+                ToolResponse response = ((ToolResponseEncoder<Object>) encoder).encode(result);
+                return response;
+            } else if (encoder instanceof ContentEncoder) {
+                if (result instanceof Iterable) {
+                    return encodeIterableWithElementEncoder((Iterable<?>) result, (ContentEncoder<Object>) encoder);
+                }
+                Content content = ((ContentEncoder<Object>) encoder).encode(result);
+                return ToolResponse.success(content);
+            } else {
+                Tr.error(tc, "CWMCM0020E.unknown.encoder.type.error");
+                throw new JSONRPCException(JSONRPCErrorCode.INTERNAL_ERROR, List.of("Unsupported encoder"));
+            }
```

Comment:
I don't think this case can happen. We don't add things which aren't `ToolResponseEncoder` or `ContentEncoder` to the encoder registry.

Therefore we can just throw an exception here with an untranslated message and comment why it shouldn't occur.